File: Bit Manipulation/singleNumberIII260.java
class Solution {
    public int[] singleNumber(int[] nums) {
        // how to split?
        // binary code, inverse code, complement code
        // 正数： 补码=反码=原码
        // 负数： 补码=反码+1
        int xor = 0;
        for (int i = 0; i < nums.length; i++) xor ^= nums[i];
        int lastDigit = xor & (-xor); 
        // to diffreciate 2 nums,  xor & (~xor + 1) also ok
        int group1 = 0, group2 = 0;
        for (int i = 0; i < nums.length; i++) {
            if ((lastDigit & nums[i]) == 0)
                group1 ^= nums[i];
            else
                group2 ^= nums[i];
        }
        return new int[]{group1, group2};
    }
}

File: Bit Manipulation/singleNumberIII260.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        bitmask = 0
        for num in nums:
            bitmask ^= num
        
        rightmost_one = bitmask & (-bitmask) # right most 1
        group1, group2 = 0, 0
        for num in nums:
            # bitmask which will contain only x
            if num & rightmost_one:
                group1 ^= num
            else:
                group2 ^= num
        return [group1, group2]